### PR TITLE
pki: generating CSR through API

### PR DESCRIPTION
**Pull request by mohammed90**
Some platforms request CSR as part of their process of generating mutually trusted credentials. This PR is to try out the idea of letting Caddy generate the CSRs as a feature of the `pki` app. The ergonomics are to be discussed.

Test run:
```shell
~ $ curl -d '{"sans": ["example.com.app"]}' -H 'Accept: application/pkcs10' http://localhost:2019/pki/ca/local/csr > my.csr

~ $ openssl req -text -noout -verify -in my.csr
Certificate request self-signature verify OK
Certificate Request:
    Data:
        Version: 1 (0x0)
        Subject:
        Subject Public Key Info:
            Public Key Algorithm: id-ecPublicKey
                Public-Key: (256 bit)
                pub:
                    04:09:65:23:b0:be:e4:2a:04:00:1f:34:72:49:a7:
                    6a:73:5b:dc:0e:68:40:57:2d:88:34:6f:57:5f:eb:
                    51:9f:aa:59:b5:28:25:b6:13:3a:b8:c3:16:7f:14:
                    47:16:ad:ac:75:ed:32:cf:88:fe:52:cf:11:91:ef:
                    2b:3a:37:ce:99
                ASN1 OID: prime256v1
                NIST CURVE: P-256
        Attributes:
            Requested Extensions:
                X509v3 Subject Alternative Name:
                    DNS:example.com.app
    Signature Algorithm: ecdsa-with-SHA256
    Signature Value:
        30:45:02:21:00:e4:8b:b3:ca:bf:35:26:04:47:85:ec:ee:05:
        65:1c:16:05:93:17:09:fe:69:41:34:5f:76:cf:66:af:c3:4b:
        f2:02:20:74:24:a0:7d:60:6b:d7:9b:8a:fb:ea:97:55:22:c3:
        d1:50:b2:e5:a9:88:9e:e1:e0:27:3f:67:31:44:d5:08:d6

~ $ cat my.csr
-----BEGIN CERTIFICATE REQUEST-----
MIHoMIGPAgEAMAAwWTATBgcqhkjOPQIBBggqhkjOPQMBBwNCAAQJZSOwvuQqBAAf
NHJJp2pzW9wOaEBXLYg0b1df61Gfqlm1KCW2Ezq4wxZ/FEcWrax17TLPiP5SzxGR
7ys6N86ZoC0wKwYJKoZIhvcNAQkOMR4wHDAaBgNVHREEEzARgg9leGFtcGxlLmNv
bS5hcHAwCgYIKoZIzj0EAwIDSAAwRQIhAOSLs8q/NSYER4Xs7gVlHBYFkxcJ/mlB
NF92z2avw0vyAiB0JKB9YGvXm4r76pdVIsPRULLlqYie4eAnP2cxRNUI1g==
-----END CERTIFICATE REQUEST-----
```

P.S.: The PR is currently a draft until the code is cleaned up and tests are added

### CODE REVIEW

**Comment by mohammed90** [comment]
On a second thought, we probably should offer the option of which private key to use: root, intermediate, or dedicated.

**Comment by mholt** [comment]
Ohh I really like where this is going! Will take a look after the weekend :)

File: modules/caddypki/adminapi.go [hunk 1]
```diff
@@ -168,6 +177,96 @@ func (a *adminAPI) handleCACerts(w http.ResponseWriter, r *http.Request) error {
...
+
+		// The curve to use with key types EC and OKP.
+		// If the Type is OKP, then acceptable curves are: Ed25519, or X25519
+		// If the Type is EC, then acceptable curves are: P-256, P-384, or P-521
+		Curve string `json:"curve,omitempty"`
+
+		// Only used with RSA keys and accepts minimum of 2048.
+		Size int `json:"size,omitempty"`
+	} `json:"key,omitempty"`
+	// SANs is a list of subject alternative names for the certificate.
```
Comment (mholt) [inline modules/caddypki/adminapi.go:204]:
```suggestion
	} `json:"key,omitempty"`

	// SANs is a list of subject alternative names for the certificate.
```

**Review by mholt — APPROVED** [review]
I like where this is going! I don't have any strong suggestions right now. I think I wouldn't do much differently.

**Comment by mohammed90** [comment]
> we probably should offer the option of which private key to use: root, intermediate, or dedicated.

I still have not addressed this because I don't know the best design for the payload to accommodate this. Ideas are welcome.

**Comment by mholt** [comment]
Maybe it could be a query string parameter?

**Comment by mholt** [comment]
@mohammed90 Just cleaning up our milestones... do you have any interest (and time, I know you are busy) to finish this up soon? If not, I wonder if closing this would relieve some pressure? And we can always revisit again later if you want to?

**Comment by mohammed90** [comment]
> @mohammed90 Just cleaning up our milestones... do you have any interest (and time, I know you are busy) to finish this up soon? If not, I wonder if closing this would relieve some pressure? And we can always revisit again later if you want to?

I don't see it important to finish. It's best we focus our energy elsewhere for now. We've figured the approach if it becomes an interest in the future.

**Pull request closed without merging**